File: sdks/python/src/fuse/_services/environments.py
from __future__ import annotations

from collections.abc import Iterator
from typing import Any
from urllib.parse import quote

from .._transport import Transport
from ..errors import ApiError
from ..types import (
    ComputerAction,
    ComputerDisplay,
    ComputerResult,
    ComputerToolResult,
    CreateRequest,
    EnvironmentInfo,
    EnvironmentPage,
    Event,
    ExecRequest,
    ExecResult,
    ForkOptions,
)
from .events import stream_events

# the server's max page size; list() requests this internally so it walks
# every page in as few round trips as possible.
_MAX_PAGE_LIMIT = 200


class EnvironmentsService:
    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def list(
        self, *, task_id: str = "", state: str = "", host_id: str = "", cursor: str = ""
    ) -> list[EnvironmentInfo]:
        # returns every environment matching the filters, transparently
        # walking every result page. for explicit single-page control (e.g.
        # a cursor from a previous call), use list_page.
        out: list[EnvironmentInfo] = []
        while True:
            page = self.list_page(
                task_id=task_id,
                state=state,
                host_id=host_id,
                limit=_MAX_PAGE_LIMIT,
                cursor=cursor,
            )
            out.extend(page.environments)
            if not page.next_cursor:
                break
            cursor = page.next_cursor
        return out

    def list_page(
        self,
        *,
        task_id: str = "",
        state: str = "",
        host_id: str = "",
        limit: int = 0,
        cursor: str = "",
    ) -> EnvironmentPage:
        # returns one page of environments matching the filters.
        params: dict[str, str] = {"task_id": task_id, "state": state, "host_id": host_id}
        if limit > 0:
            params["limit"] = str(limit)
        if cursor:
            params["cursor"] = cursor
        resp = self._t.request("GET", "/v1/environments", params=params)
        return EnvironmentPage.model_validate(resp.json())

    def get(self, vm_id: str) -> EnvironmentInfo:
        if not vm_id:
            raise ValueError("vm id is required")
        resp = self._t.request("GET", f"/v1/environments/{quote(vm_id, safe='')}")
        return EnvironmentInfo.model_validate(resp.json())

    def create(self, request: CreateRequest) -> EnvironmentInfo:
        resp = self._t.request("POST", "/v1/environments", body=request)
        return EnvironmentInfo.model_validate(resp.json())

    def drain(self, vm_id: str) -> EnvironmentInfo:
        # phase-1 only: this quiesces the guest workload and moves the vm to
        # "draining" so you can inspect its output, but it does not destroy
        # the vm. call environments.destroy(vm_id) afterward to remove it -
        # a drained vm left undestroyed keeps running (and billing) forever.
        return self._action(vm_id, "drain")

    def fork(
        self, vm_id: str, options: ForkOptions | None = None
    ) -> EnvironmentInfo:
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        resp = self._t.request(
            "POST", path, params={"action": "fork"}, body=options or ForkOptions()
        )
        return EnvironmentInfo.model_validate(resp.json())

    def computer(self, vm_id: str, action: ComputerAction) -> ComputerResult:
        # relays one computer-use action to the environment's desktop and
        # returns the result, usually carrying a base64 png screenshot.
        # requires an environment booted from a desktop image; on any other
        # image the server answers 503 with a reason.
        if not vm_id:
            raise ValueError("vm id is required")
        if not action.action:
            raise ValueError("action is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}/computer"
        resp = self._t.request("POST", path, body=action)
        return ComputerResult.model_validate(resp.json())

    def computer_tool_result(
        self, vm_id: str, tool_input: dict[str, Any]
    ) -> ComputerToolResult:
        # executes one computer tool_use against the environment and shapes
        # the answer for the messages api. tool_input is the tool_use block's
        # input, verbatim: it is already the wire shape the computer endpoint
        # takes, and it is forwarded untouched so action fields this sdk has
        # not learned yet still reach the guest.
        #
        # an action the environment refuses (a malformed coordinate, a
        # display that is not up) comes back as is_error content rather than
        # raising: the agent loop should feed it to the model, which is the
        # party that can correct it. a raise is reserved for failures the
        # model cannot fix - transport, auth, an unknown environment.
        if not vm_id:
            raise ValueError("vm id is required")
        if not tool_input.get("action"):
            return ComputerToolResult(
                content=[{"type": "text", "text": "tool input names no action"}],
                is_error=True,
            )
        path = f"/v1/environments/{quote(vm_id, safe='')}/computer"
        try:
            resp = self._t.request("POST", path, body=tool_input)
        except ApiError as err:
            if err.status in (400, 409, 503):
                return ComputerToolResult(
                    content=[
                        {
                            "type": "text",
                            "text": err.message or "the computer action failed",
                        }
                    ],
                    is_error=True,
                )
            raise
        data = resp.json()
        content: list[dict[str, Any]] = []
        if data.get("output"):
            content.append({"type": "text", "text": data["output"]})
        if data.get("screenshot"):
            content.append(
                {
                    "type": "image",
                    "source": {
                        "type": "base64",
                        "media_type": "image/png",
                        "data": data["screenshot"],
                    },
                }
            )
        if not content:
            content.append({"type": "text", "text": "done"})
        return ComputerToolResult(content=content)

    def computer_display(self, vm_id: str) -> ComputerDisplay:
        # reports whether the environment has a live display and at what
        # geometry. up is false with a reason on an image with no desktop.
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}/computer"
        resp = self._t.request("GET", path)
        return ComputerDisplay.model_validate(resp.json())

    def exec(self, vm_id: str, request: ExecRequest) -> ExecResult:
        # runs a command inside a running environment's guest. requires the
        # master token.
        #
        # a non-zero exit_code is returned, not raised: the command ran and
        # failed. an ApiError means the command could not be run at all.
        if not vm_id:
            raise ValueError("vm id is required")
        has_cmd = bool(request.cmd)
        has_shell = bool(request.shell)
        if not has_cmd and not has_shell:
            raise ValueError("one of cmd or shell is required")
        if has_cmd and has_shell:
            raise ValueError("cmd and shell are mutually exclusive")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        resp = self._t.request("POST", path, params={"action": "exec"}, body=request)
        return ExecResult.model_validate(resp.json())

    def rotate_token(self, vm_id: str) -> None:
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        self._t.request("POST", path, params={"action": "rotate-token"})

    def destroy(self, vm_id: str) -> None:
        if not vm_id:
            raise ValueError("vm id is required")
        self._t.request("DELETE", f"/v1/environments/{quote(vm_id, safe='')}")

    def events(self, vm_id: str) -> Iterator[Event]:
        # opens the sse stream and yields Event values. the iterator ends
        # cleanly on eof, after a terminal-state event, or after a final
        # Event whose err is set on a stream-level failure.
        return stream_events(self._t, vm_id)

    def _action(self, vm_id: str, action: str) -> EnvironmentInfo:
        if not vm_id:
            raise ValueError("vm id is required")
        if not action:
            raise ValueError("action is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        resp = self._t.request("POST", path, params={"action": action})
        return EnvironmentInfo.model_validate(resp.json())

File: sdks/rust/src/environments.rs
use std::sync::Arc;
use std::time::Duration;

use reqwest::Method;
use serde::ser::SerializeStruct;
use serde::{Deserialize, Serialize, Serializer};

use crate::error::Error;
use crate::events::EventStream;
use crate::transport::{pagination, require, Transport, MAX_PAGE_LIMIT};
use crate::types::{CreateRequest, EnvironmentInfo, EnvironmentState, ForkOptions};

// mirrors the orchestrator's own default guest-command bound. it is what the
// server applies when the timeout is omitted, so the client waits at least
// that long before giving up.
const DEFAULT_EXEC_SERVER_TIMEOUT: Duration = Duration::from_secs(60);

// headroom on top of the guest timeout for the hops the guest timeout does
// not cover: the ssh connect on the host and the network round trip. without
// it a command that runs for exactly its guest timeout would race the
// client's deadline.
const EXEC_HTTP_OVERHEAD: Duration = Duration::from_secs(30);

/// Filters and pagination for [`Environments::list`] and
/// [`Environments::list_page`].
#[derive(Debug, Clone, Default)]
#[non_exhaustive]
pub struct ListEnvironmentsOptions {
    pub task_id: Option<String>,
    pub state: Option<EnvironmentState>,
    pub host_id: Option<String>,
    /// Caps the page size (server default 50, max 200). Only consulted by
    /// [`Environments::list_page`] — [`Environments::list`] always requests
    /// the server's max page size internally since it walks every page.
    pub limit: Option<u32>,
    /// Resumes after a previous page's [`EnvironmentPage::next_cursor`]. It
    /// is an opaque token; treat it as such rather than parsing it.
    pub cursor: Option<String>,
}

impl ListEnvironmentsOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn task_id(mut self, task_id: impl Into<String>) -> Self {
        self.task_id = Some(task_id.into());
        self
    }

    pub fn state(mut self, state: EnvironmentState) -> Self {
        self.state = Some(state);
        self
    }

    pub fn host_id(mut self, host_id: impl Into<String>) -> Self {
        self.host_id = Some(host_id.into());
        self
    }

    pub fn limit(mut self, limit: u32) -> Self {
        self.limit = Some(limit);
        self
    }

    pub fn cursor(mut self, cursor: impl Into<String>) -> Self {
        self.cursor = Some(cursor.into());
        self
    }
}

/// One page of a list call, plus the cursor to fetch the next one.
#[derive(Debug, Clone, Default, Deserialize)]
pub struct EnvironmentPage {
    #[serde(default)]
    pub environments: Vec<EnvironmentInfo>,
    /// `None` once there are no more results.
    #[serde(default)]
    pub next_cursor: Option<String>,
}

/// The command half of an [`ExecRequest`]. Exactly one of argv and shell
/// goes on the wire, which this enum makes impossible to get wrong.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExecCommand {
    /// An argv run directly in the guest, e.g. `["ls", "-l"]`. Argv needs
    /// no quoting rules and cannot be turned into an injection by
    /// interpolating a value, so prefer it.
    Argv(Vec<String>),
    /// A string run under `sh -lc`. Use it only for what argv cannot
    /// express: pipelines, redirects, and globs.
    Shell(String),
}

/// The body of [`Environments::exec`].
///
/// ```
/// use std::time::Duration;
/// use fuse::ExecRequest;
///
/// let ls = ExecRequest::cmd(["ls", "-l", "/tmp"]);
/// let pipeline = ExecRequest::shell("dmesg | tail -n 5").timeout(Duration::from_secs(10));
/// ```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExecRequest {
    pub command: ExecCommand,
    /// Bounds the command inside the guest, in milliseconds. `None` takes
    /// the server default; the server clamps anything above its ceiling.
    pub timeout_ms: Option<u64>,
}

impl ExecRequest {
    /// Returns a request running `argv` directly in the guest.
    pub fn cmd<I, S>(argv: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        Self {
            command: ExecCommand::Argv(argv.into_iter().map(Into::into).collect()),
            timeout_ms: None,
        }
    }

    /// Returns a request running `script` under `sh -lc`.
    pub fn shell(script: impl Into<String>) -> Self {
        Self {
            command: ExecCommand::Shell(script.into()),
            timeout_ms: None,
        }
    }

    /// Bounds the command inside the guest. Sub-millisecond precision is
    /// truncated.
    pub fn timeout(mut self, timeout: Duration) -> Self {
        self.timeout_ms = Some(timeout.as_millis().try_into().unwrap_or(u64::MAX));
        self
    }

    pub fn timeout_ms(mut self, timeout_ms: u64) -> Self {
        self.timeout_ms = Some(timeout_ms);
        self
    }
}

impl Serialize for ExecRequest {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut body = serializer.serialize_struct("ExecRequest", 2)?;
        match &self.command {
            ExecCommand::Argv(argv) => body.serialize_field("cmd", argv)?,
            ExecCommand::Shell(script) => body.serialize_field("shell", script)?,
        }
        match self.timeout_ms {
            Some(timeout_ms) => body.serialize_field("timeout_ms", &timeout_ms)?,
            None => body.skip_field("timeout_ms")?,
        }
        body.end()
    }
}

/// The outcome of a guest command.
///
/// A non-zero exit code is a successful call: the command ran and failed.
/// Only an `Err` from [`Environments::exec`] means the command could not be
/// run at all.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct ExecResult {
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
}

impl ExecResult {
    /// Reports whether the command exited zero.
    pub fn success(&self) -> bool {
        self.exit_code == 0
    }
}

/// The environments service, from
/// [`Client::environments`](crate::Client::environments).
#[derive(Clone)]
pub struct Environments {
    t: Arc<Transport>,
}

impl Environments {
    pub(crate) fn new(t: Arc<Transport>) -> Self {
        Self { t }
    }

    pub(crate) fn transport(&self) -> &Transport {
        &self.t
    }

    /// Returns every environment matching `options`, transparently walking
    /// every result page. For explicit single-page control (e.g. a CLI
    /// `--cursor` flag), use [`Environments::list_page`].
    pub async fn list(
        &self,
        options: ListEnvironmentsOptions,
    ) -> Result<Vec<EnvironmentInfo>, Error> {
        let mut out = Vec::new();
        let mut page_options = options;
        page_options.limit = Some(MAX_PAGE_LIMIT);
        loop {
            let page = self.list_page(page_options.clone()).await?;
            out.extend(page.environments);
            match page.next_cursor {
                Some(cursor) if !cursor.is_empty() => page_options.cursor = Some(cursor),
                _ => return Ok(out),
            }
        }
    }

    /// Returns one page of environments matching `options`.
    pub async fn list_page(
        &self,
        options: ListEnvironmentsOptions,
    ) -> Result<EnvironmentPage, Error> {
        let mut query: Vec<(&str, String)> = Vec::new();
        if let Some(task_id) = &options.task_id {
            query.push(("task_id", task_id.clone()));
        }
        if let Some(state) = &options.state {
            query.push(("state", state.to_string()));
        }
        if let Some(host_id) = &options.host_id {
            query.push(("host_id", host_id.clone()));
        }
        pagination(&mut query, options.limit, &options.cursor);
        let request = self
            .t
            .request(Method::GET, &["v1", "environments"])
            .query(&query);
        self.t.send_json(request).await
    }

    /// Returns one environment by id.
    pub async fn get(&self, vm_id: &str) -> Result<EnvironmentInfo, Error> {
        require(vm_id, "vm id")?;
        let request = self.t.request(Method::GET, &["v1", "environments", vm_id]);
        self.t.send_json(request).await
    }

    /// Creates an environment. The call returns as soon as the VM is up;
    /// follow [`Environments::events`] to watch it settle.
    pub async fn create(&self, request: CreateRequest) -> Result<EnvironmentInfo, Error> {
        let request = self
            .t
            .request(Method::POST, &["v1", "environments"])
            .json(&request);
        self.t.send_json(request).await
    }

    /// Starts draining an environment.
    pub async fn drain(&self, vm_id: &str) -> Result<EnvironmentInfo, Error> {
        self.action(vm_id, "drain").await
    }

    /// Creates a new environment seeded from an existing one and returns
    /// the new environment. When `options.reuse_snapshot_id` is `None` the
    /// server snapshots the source first; otherwise it reuses the named
    /// snapshot.
    pub async fn fork(&self, vm_id: &str, options: ForkOptions) -> Result<EnvironmentInfo, Error> {
        require(vm_id, "vm id")?;
        let request = self
            .t
            .request(Method::POST, &["v1", "environments", vm_id])
            .query(&[("action", "fork")])
            .json(&options);
        self.t.send_json(request).await
    }

    /// Rotates the environment's own credentials.
    pub async fn rotate_token(&self, vm_id: &str) -> Result<(), Error> {
        require(vm_id, "vm id")?;
        let request = self
            .t
            .request(Method::POST, &["v1", "environments", vm_id])
            .query(&[("action", "rotate-token")]);
        self.t.send_unit(request).await
    }

    /// Destroys an environment.
    pub async fn destroy(&self, vm_id: &str) -> Result<(), Error> {
        require(vm_id, "vm id")?;
        let request = self
            .t
            .request(Method::DELETE, &["v1", "environments", vm_id]);
        self.t.send_unit(request).await
    }

    /// Runs a command inside a running environment's guest and returns its
    /// exit code with stdout and stderr kept separate.
    ///
    /// Exec requires the master token.
    pub async fn exec(&self, vm_id: &str, request: ExecRequest) -> Result<ExecResult, Error> {
        require(vm_id, "vm id")?;
        match &request.command {
            ExecCommand::Argv(argv) if argv.is_empty() => {
                return Err(Error::InvalidRequest("cmd must not be empty".to_owned()));
            }
            ExecCommand::Shell(script) if script.is_empty() => {
                return Err(Error::InvalidRequest("shell must not be empty".to_owned()));
            }
            _ => {}
        }

        // a guest command can run far longer than the default client
        // timeout, so this uses the no-timeout stream client and bounds the
        // call by a per-request deadline derived from the requested guest
        // timeout instead. without this the 60s client default would cut a
        // longer command short awaiting headers.
        let guest = request
            .timeout_ms
            .filter(|&ms| ms > 0)
            .map(Duration::from_millis)
            .unwrap_or(DEFAULT_EXEC_SERVER_TIMEOUT);
        let http_request = self
            .t
            .stream_request(Method::POST, &["v1", "environments", vm_id])
            .query(&[("action", "exec")])
            .timeout(guest + EXEC_HTTP_OVERHEAD)
            .json(&request);
        self.t.send_json(http_request).await
    }

    /// Opens the SSE event stream for an environment. The stream ends
    /// cleanly after a terminal state event (destroyed/failed) or on EOF;
    /// drop it to stop early.
    pub async fn events(&self, vm_id: &str) -> Result<EventStream, Error> {
        require(vm_id, "vm id")?;
        let response = self
            .t
            .stream_request(Method::GET, &["v1", "environments", vm_id, "events"])
            .header(reqwest::header::ACCEPT, "text/event-stream")
            .send()
            .await?;
        let response = Transport::check(response).await?;
        Ok(EventStream::new(response))
    }

    async fn action(&self, vm_id: &str, action: &str) -> Result<EnvironmentInfo, Error> {
        require(vm_id, "vm id")?;
        let request = self
            .t
            .request(Method::POST, &["v1", "environments", vm_id])
            .query(&[("action", action)]);
        self.t.send_json(request).await
    }
}
